{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "gpuType": "T4"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "LKUaNbBPhINz",
        "outputId": "28bb5062-4333-4318-e4e6-5aee80c290fb"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'protein_bert'...\n",
            "remote: Enumerating objects: 249, done.\u001b[K\n",
            "remote: Counting objects: 100% (97/97), done.\u001b[K\n",
            "remote: Compressing objects: 100% (36/36), done.\u001b[K\n",
            "remote: Total 249 (delta 83), reused 61 (delta 61), pack-reused 152 (from 1)\u001b[K\n",
            "Receiving objects: 100% (249/249), 23.44 MiB | 19.43 MiB/s, done.\n",
            "Resolving deltas: 100% (124/124), done.\n"
          ]
        }
      ],
      "source": [
        "# Clone the ProteinBERT repository\n",
        "!git clone https://github.com/nadavbra/protein_bert.git"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "%cd protein_bert"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "K7bCkKcjhhfx",
        "outputId": "98ec6225-584b-4163-b897-31f9e139686e"
      },
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/content/protein_bert\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "3qh533eViICE"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Initialize and update the submodules\n",
        "!git submodule init\n",
        "!git submodule update"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NnxQAoLAhjvS",
        "outputId": "10bac877-2953-4034-b776-0a420d3661d7"
      },
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Submodule 'proteinbert/shared_utils' (https://github.com/nadavbra/shared_utils.git) registered for path 'proteinbert/shared_utils'\n",
            "Cloning into '/content/protein_bert/proteinbert/shared_utils'...\n",
            "Submodule path 'proteinbert/shared_utils': checked out 'dc1c62a1754c51f6d46b7486e4a3e5e62c0570e1'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Install ProteinBERT and dependencies\n",
        "!python setup.py install"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3xB_gEZzhmtv",
        "outputId": "a054a78f-1c90-4b9c-e840-fef0adab0229"
      },
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "running install\n",
            "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/cmd.py:66: SetuptoolsDeprecationWarning: setup.py install is deprecated.\n",
            "!!\n",
            "\n",
            "        ********************************************************************************\n",
            "        Please avoid running ``setup.py`` directly.\n",
            "        Instead, use pypa/build, pypa/installer or other\n",
            "        standards-based tools.\n",
            "\n",
            "        See https://blog.ganssle.io/articles/2021/10/setup-py-deprecated.html for details.\n",
            "        ********************************************************************************\n",
            "\n",
            "!!\n",
            "  self.initialize_options()\n",
            "/usr/local/lib/python3.10/dist-packages/setuptools/_distutils/cmd.py:66: EasyInstallDeprecationWarning: easy_install command is deprecated.\n",
            "!!\n",
            "\n",
            "        ********************************************************************************\n",
            "        Please avoid running ``setup.py`` and ``easy_install``.\n",
            "        Instead, use pypa/build, pypa/installer or other\n",
            "        standards-based tools.\n",
            "\n",
            "        See https://github.com/pypa/setuptools/issues/917 for details.\n",
            "        ********************************************************************************\n",
            "\n",
            "!!\n",
            "  self.initialize_options()\n",
            "running bdist_egg\n",
            "running egg_info\n",
            "creating protein_bert.egg-info\n",
            "writing protein_bert.egg-info/PKG-INFO\n",
            "writing dependency_links to protein_bert.egg-info/dependency_links.txt\n",
            "writing requirements to protein_bert.egg-info/requires.txt\n",
            "writing top-level names to protein_bert.egg-info/top_level.txt\n",
            "writing manifest file 'protein_bert.egg-info/SOURCES.txt'\n",
            "reading manifest file 'protein_bert.egg-info/SOURCES.txt'\n",
            "writing manifest file 'protein_bert.egg-info/SOURCES.txt'\n",
            "installing library code to build/bdist.linux-x86_64/egg\n",
            "running install_lib\n",
            "running build_py\n",
            "creating build/lib/proteinbert\n",
            "copying proteinbert/finetuning.py -> build/lib/proteinbert\n",
            "copying proteinbert/tokenization.py -> build/lib/proteinbert\n",
            "copying proteinbert/conv_and_global_attention_model.py -> build/lib/proteinbert\n",
            "copying proteinbert/__init__.py -> build/lib/proteinbert\n",
            "copying proteinbert/uniref_dataset.py -> build/lib/proteinbert\n",
            "copying proteinbert/pretraining.py -> build/lib/proteinbert\n",
            "copying proteinbert/model_generation.py -> build/lib/proteinbert\n",
            "copying proteinbert/existing_model_loading.py -> build/lib/proteinbert\n",
            "creating build/lib/proteinbert/shared_utils\n",
            "copying proteinbert/shared_utils/reference_genome.py -> build/lib/proteinbert/shared_utils\n",
            "copying proteinbert/shared_utils/__init__.py -> build/lib/proteinbert/shared_utils\n",
            "copying proteinbert/shared_utils/util.py -> build/lib/proteinbert/shared_utils\n",
            "creating build/bdist.linux-x86_64/egg\n",
            "creating build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/finetuning.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/tokenization.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/conv_and_global_attention_model.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/__init__.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/uniref_dataset.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/pretraining.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "creating build/bdist.linux-x86_64/egg/proteinbert/shared_utils\n",
            "copying build/lib/proteinbert/shared_utils/reference_genome.py -> build/bdist.linux-x86_64/egg/proteinbert/shared_utils\n",
            "copying build/lib/proteinbert/shared_utils/__init__.py -> build/bdist.linux-x86_64/egg/proteinbert/shared_utils\n",
            "copying build/lib/proteinbert/shared_utils/util.py -> build/bdist.linux-x86_64/egg/proteinbert/shared_utils\n",
            "copying build/lib/proteinbert/model_generation.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "copying build/lib/proteinbert/existing_model_loading.py -> build/bdist.linux-x86_64/egg/proteinbert\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/finetuning.py to finetuning.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/tokenization.py to tokenization.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/conv_and_global_attention_model.py to conv_and_global_attention_model.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/__init__.py to __init__.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/uniref_dataset.py to uniref_dataset.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/pretraining.py to pretraining.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/shared_utils/reference_genome.py to reference_genome.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/shared_utils/__init__.py to __init__.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/shared_utils/util.py to util.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/model_generation.py to model_generation.cpython-310.pyc\n",
            "byte-compiling build/bdist.linux-x86_64/egg/proteinbert/existing_model_loading.py to existing_model_loading.cpython-310.pyc\n",
            "creating build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "installing scripts to build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "running install_scripts\n",
            "running build_scripts\n",
            "creating build/scripts-3.10\n",
            "copying and adjusting bin/create_uniref_db -> build/scripts-3.10\n",
            "copying and adjusting bin/create_uniref_h5_dataset -> build/scripts-3.10\n",
            "copying and adjusting bin/pretrain_proteinbert -> build/scripts-3.10\n",
            "copying and adjusting bin/set_h5_testset -> build/scripts-3.10\n",
            "changing mode of build/scripts-3.10/create_uniref_db from 644 to 755\n",
            "changing mode of build/scripts-3.10/create_uniref_h5_dataset from 644 to 755\n",
            "changing mode of build/scripts-3.10/pretrain_proteinbert from 644 to 755\n",
            "changing mode of build/scripts-3.10/set_h5_testset from 644 to 755\n",
            "creating build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "copying build/scripts-3.10/pretrain_proteinbert -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "copying build/scripts-3.10/create_uniref_db -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "copying build/scripts-3.10/create_uniref_h5_dataset -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "copying build/scripts-3.10/set_h5_testset -> build/bdist.linux-x86_64/egg/EGG-INFO/scripts\n",
            "changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/pretrain_proteinbert to 755\n",
            "changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/create_uniref_db to 755\n",
            "changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/create_uniref_h5_dataset to 755\n",
            "changing mode of build/bdist.linux-x86_64/egg/EGG-INFO/scripts/set_h5_testset to 755\n",
            "copying protein_bert.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying protein_bert.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying protein_bert.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying protein_bert.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "copying protein_bert.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO\n",
            "zip_safe flag not set; analyzing archive contents...\n",
            "creating dist\n",
            "creating 'dist/protein_bert-1.0.1-py3.10.egg' and adding 'build/bdist.linux-x86_64/egg' to it\n",
            "removing 'build/bdist.linux-x86_64/egg' (and everything under it)\n",
            "Processing protein_bert-1.0.1-py3.10.egg\n",
            "Copying protein_bert-1.0.1-py3.10.egg to /usr/local/lib/python3.10/dist-packages\n",
            "Adding protein-bert 1.0.1 to easy-install.pth file\n",
            "Installing create_uniref_db script to /usr/local/bin\n",
            "Installing create_uniref_h5_dataset script to /usr/local/bin\n",
            "Installing pretrain_proteinbert script to /usr/local/bin\n",
            "Installing set_h5_testset script to /usr/local/bin\n",
            "\n",
            "Installed /usr/local/lib/python3.10/dist-packages/protein_bert-1.0.1-py3.10.egg\n",
            "Processing dependencies for protein-bert==1.0.1\n",
            "Searching for pyfaidx\n",
            "Reading https://pypi.org/simple/pyfaidx/\n",
            "Downloading https://files.pythonhosted.org/packages/38/1c/74a91b91db35b16fff976f52e81a15b8836240ed5aeeb5d1c7913038efff/pyfaidx-0.8.1.3-py3-none-any.whl#sha256=044abb9b7b03fef1cc393ae6831ece80df5c1a4dff295fa33808c6f63a81cdc4\n",
            "Best match: pyfaidx 0.8.1.3\n",
            "Processing pyfaidx-0.8.1.3-py3-none-any.whl\n",
            "Installing pyfaidx-0.8.1.3-py3-none-any.whl to /usr/local/lib/python3.10/dist-packages\n",
            "Adding pyfaidx 0.8.1.3 to easy-install.pth file\n",
            "detected new path './protein_bert-1.0.1-py3.10.egg'\n",
            "Installing faidx script to /usr/local/bin\n",
            "\n",
            "Installed /usr/local/lib/python3.10/dist-packages/pyfaidx-0.8.1.3-py3.10.egg\n",
            "Searching for tensorflow_addons\n",
            "Reading https://pypi.org/simple/tensorflow_addons/\n",
            "Downloading https://files.pythonhosted.org/packages/c0/a0/2bb2edb863555eb87e8ed45ded678138f454743a6ee90e0ee994f00a7147/tensorflow_addons-0.23.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl#sha256=0d224e237c11a0185fe1ea5d72e52c9d183a3adc075b4684d904f9b80b4c2579\n",
            "Best match: tensorflow-addons 0.23.0\n",
            "Processing tensorflow_addons-0.23.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl\n",
            "Installing tensorflow_addons-0.23.0-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl to /usr/local/lib/python3.10/dist-packages\n",
            "Adding tensorflow-addons 0.23.0 to easy-install.pth file\n",
            "detected new path './pyfaidx-0.8.1.3-py3.10.egg'\n",
            "\n",
            "Installed /usr/local/lib/python3.10/dist-packages/tensorflow_addons-0.23.0-py3.10-linux-x86_64.egg\n",
            "Searching for typeguard<3.0.0,>=2.7\n",
            "Reading https://pypi.org/simple/typeguard/\n",
            "Downloading https://files.pythonhosted.org/packages/9a/bb/d43e5c75054e53efce310e79d63df0ac3f25e34c926be5dffb7d283fb2a8/typeguard-2.13.3-py3-none-any.whl#sha256=5e3e3be01e887e7eafae5af63d1f36c849aaa94e3a0112097312aabfa16284f1\n",
            "Best match: typeguard 2.13.3\n",
            "Processing typeguard-2.13.3-py3-none-any.whl\n",
            "Installing typeguard-2.13.3-py3-none-any.whl to /usr/local/lib/python3.10/dist-packages\n",
            "Adding typeguard 2.13.3 to easy-install.pth file\n",
            "detected new path './tensorflow_addons-0.23.0-py3.10-linux-x86_64.egg'\n",
            "\n",
            "Installed /usr/local/lib/python3.10/dist-packages/typeguard-2.13.3-py3.10.egg\n",
            "Searching for lxml==5.3.0\n",
            "Best match: lxml 5.3.0\n",
            "Adding lxml 5.3.0 to easy-install.pth file\n",
            "detected new path './typeguard-2.13.3-py3.10.egg'\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for h5py==3.12.1\n",
            "Best match: h5py 3.12.1\n",
            "Adding h5py 3.12.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for pandas==2.2.2\n",
            "Best match: pandas 2.2.2\n",
            "Adding pandas 2.2.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for numpy==1.26.4\n",
            "Best match: numpy 1.26.4\n",
            "Adding numpy 1.26.4 to easy-install.pth file\n",
            "Installing f2py script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for tensorflow==2.17.1\n",
            "Best match: tensorflow 2.17.1\n",
            "Adding tensorflow 2.17.1 to easy-install.pth file\n",
            "Installing import_pb_to_tensorboard script to /usr/local/bin\n",
            "Installing saved_model_cli script to /usr/local/bin\n",
            "Installing tensorboard script to /usr/local/bin\n",
            "Installing tf_upgrade_v2 script to /usr/local/bin\n",
            "Installing tflite_convert script to /usr/local/bin\n",
            "Installing toco script to /usr/local/bin\n",
            "Installing toco_from_protos script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for packaging==24.2\n",
            "Best match: packaging 24.2\n",
            "Adding packaging 24.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for importlib-metadata==8.5.0\n",
            "Best match: importlib-metadata 8.5.0\n",
            "Adding importlib-metadata 8.5.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for tzdata==2024.2\n",
            "Best match: tzdata 2024.2\n",
            "Adding tzdata 2024.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for pytz==2024.2\n",
            "Best match: pytz 2024.2\n",
            "Adding pytz 2024.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for python-dateutil==2.8.2\n",
            "Best match: python-dateutil 2.8.2\n",
            "Adding python-dateutil 2.8.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for tensorflow-io-gcs-filesystem==0.37.1\n",
            "Best match: tensorflow-io-gcs-filesystem 0.37.1\n",
            "Adding tensorflow-io-gcs-filesystem 0.37.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for keras==3.5.0\n",
            "Best match: keras 3.5.0\n",
            "Adding keras 3.5.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for tensorboard==2.17.1\n",
            "Best match: tensorboard 2.17.1\n",
            "Adding tensorboard 2.17.1 to easy-install.pth file\n",
            "Installing tensorboard script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for grpcio==1.68.0\n",
            "Best match: grpcio 1.68.0\n",
            "Adding grpcio 1.68.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for wrapt==1.16.0\n",
            "Best match: wrapt 1.16.0\n",
            "Adding wrapt 1.16.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for typing-extensions==4.12.2\n",
            "Best match: typing-extensions 4.12.2\n",
            "Adding typing-extensions 4.12.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages/setuptools/_vendor\n",
            "Searching for termcolor==2.5.0\n",
            "Best match: termcolor 2.5.0\n",
            "Adding termcolor 2.5.0 to easy-install.pth file\n",
            "detected new path './setuptools/_vendor'\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for six==1.16.0\n",
            "Best match: six 1.16.0\n",
            "Adding six 1.16.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for setuptools==75.1.0\n",
            "Best match: setuptools 75.1.0\n",
            "Adding setuptools 75.1.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for requests==2.32.3\n",
            "Best match: requests 2.32.3\n",
            "Adding requests 2.32.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for protobuf==4.25.5\n",
            "Best match: protobuf 4.25.5\n",
            "Adding protobuf 4.25.5 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for opt-einsum==3.4.0\n",
            "Best match: opt-einsum 3.4.0\n",
            "Adding opt-einsum 3.4.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for ml-dtypes==0.4.1\n",
            "Best match: ml-dtypes 0.4.1\n",
            "Adding ml-dtypes 0.4.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for libclang==18.1.1\n",
            "Best match: libclang 18.1.1\n",
            "Adding libclang 18.1.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for google-pasta==0.2.0\n",
            "Best match: google-pasta 0.2.0\n",
            "Adding google-pasta 0.2.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for gast==0.6.0\n",
            "Best match: gast 0.6.0\n",
            "Adding gast 0.6.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for flatbuffers==24.3.25\n",
            "Best match: flatbuffers 24.3.25\n",
            "Adding flatbuffers 24.3.25 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for astunparse==1.6.3\n",
            "Best match: astunparse 1.6.3\n",
            "Adding astunparse 1.6.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for absl-py==1.4.0\n",
            "Best match: absl-py 1.4.0\n",
            "Adding absl-py 1.4.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for zipp==3.21.0\n",
            "Best match: zipp 3.21.0\n",
            "Adding zipp 3.21.0 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for optree==0.13.1\n",
            "Best match: optree 0.13.1\n",
            "Adding optree 0.13.1 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for namex==0.0.8\n",
            "Best match: namex 0.0.8\n",
            "Adding namex 0.0.8 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for rich==13.9.4\n",
            "Best match: rich 13.9.4\n",
            "Adding rich 13.9.4 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for werkzeug==3.1.3\n",
            "Best match: werkzeug 3.1.3\n",
            "Adding werkzeug 3.1.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for tensorboard-data-server==0.7.2\n",
            "Best match: tensorboard-data-server 0.7.2\n",
            "Adding tensorboard-data-server 0.7.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for Markdown==3.7\n",
            "Best match: Markdown 3.7\n",
            "Adding Markdown 3.7 to easy-install.pth file\n",
            "Installing markdown_py script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for certifi==2024.8.30\n",
            "Best match: certifi 2024.8.30\n",
            "Adding certifi 2024.8.30 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for urllib3==2.2.3\n",
            "Best match: urllib3 2.2.3\n",
            "Adding urllib3 2.2.3 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for idna==3.10\n",
            "Best match: idna 3.10\n",
            "Adding idna 3.10 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for charset-normalizer==3.4.0\n",
            "Best match: charset-normalizer 3.4.0\n",
            "Adding charset-normalizer 3.4.0 to easy-install.pth file\n",
            "Installing normalizer script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for wheel==0.45.0\n",
            "Best match: wheel 0.45.0\n",
            "Adding wheel 0.45.0 to easy-install.pth file\n",
            "Installing wheel script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for pygments==2.18.0\n",
            "Best match: pygments 2.18.0\n",
            "Adding pygments 2.18.0 to easy-install.pth file\n",
            "Installing pygmentize script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for markdown-it-py==3.0.0\n",
            "Best match: markdown-it-py 3.0.0\n",
            "Adding markdown-it-py 3.0.0 to easy-install.pth file\n",
            "Installing markdown-it script to /usr/local/bin\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for MarkupSafe==3.0.2\n",
            "Best match: MarkupSafe 3.0.2\n",
            "Adding MarkupSafe 3.0.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Searching for mdurl==0.1.2\n",
            "Best match: mdurl 0.1.2\n",
            "Adding mdurl 0.1.2 to easy-install.pth file\n",
            "\n",
            "Using /usr/local/lib/python3.10/dist-packages\n",
            "Finished processing dependencies for protein-bert==1.0.1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from proteinbert import load_pretrained_model\n",
        "\n",
        "model, input_encoder = load_pretrained_model()"
      ],
      "metadata": {
        "id": "PawbTSmGiKQ6"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!pip install biopython tensorflow"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "jHOLrQ6divS8",
        "outputId": "6c55d420-e742-43ef-e749-d9cedfb75623"
      },
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting biopython\n",
            "  Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (12 kB)\n",
            "Requirement already satisfied: tensorflow in /usr/local/lib/python3.10/dist-packages (2.17.1)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n",
            "Requirement already satisfied: absl-py>=1.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.4.0)\n",
            "Requirement already satisfied: astunparse>=1.6.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.6.3)\n",
            "Requirement already satisfied: flatbuffers>=24.3.25 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (24.3.25)\n",
            "Requirement already satisfied: gast!=0.5.0,!=0.5.1,!=0.5.2,>=0.2.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.6.0)\n",
            "Requirement already satisfied: google-pasta>=0.1.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.2.0)\n",
            "Requirement already satisfied: h5py>=3.10.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.12.1)\n",
            "Requirement already satisfied: libclang>=13.0.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (18.1.1)\n",
            "Requirement already satisfied: ml-dtypes<0.5.0,>=0.3.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.4.1)\n",
            "Requirement already satisfied: opt-einsum>=2.3.2 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.4.0)\n",
            "Requirement already satisfied: packaging in /usr/local/lib/python3.10/dist-packages (from tensorflow) (24.2)\n",
            "Requirement already satisfied: protobuf!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.20.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (4.25.5)\n",
            "Requirement already satisfied: requests<3,>=2.21.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.32.3)\n",
            "Requirement already satisfied: setuptools in /usr/local/lib/python3.10/dist-packages (from tensorflow) (75.1.0)\n",
            "Requirement already satisfied: six>=1.12.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.16.0)\n",
            "Requirement already satisfied: termcolor>=1.1.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.5.0)\n",
            "Requirement already satisfied: typing-extensions>=3.6.6 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (4.12.2)\n",
            "Requirement already satisfied: wrapt>=1.11.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.16.0)\n",
            "Requirement already satisfied: grpcio<2.0,>=1.24.3 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (1.68.0)\n",
            "Requirement already satisfied: tensorboard<2.18,>=2.17 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (2.17.1)\n",
            "Requirement already satisfied: keras>=3.2.0 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (3.5.0)\n",
            "Requirement already satisfied: tensorflow-io-gcs-filesystem>=0.23.1 in /usr/local/lib/python3.10/dist-packages (from tensorflow) (0.37.1)\n",
            "Requirement already satisfied: wheel<1.0,>=0.23.0 in /usr/local/lib/python3.10/dist-packages (from astunparse>=1.6.0->tensorflow) (0.45.0)\n",
            "Requirement already satisfied: rich in /usr/local/lib/python3.10/dist-packages (from keras>=3.2.0->tensorflow) (13.9.4)\n",
            "Requirement already satisfied: namex in /usr/local/lib/python3.10/dist-packages (from keras>=3.2.0->tensorflow) (0.0.8)\n",
            "Requirement already satisfied: optree in /usr/local/lib/python3.10/dist-packages (from keras>=3.2.0->tensorflow) (0.13.1)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorflow) (3.4.0)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorflow) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorflow) (2.2.3)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.10/dist-packages (from requests<3,>=2.21.0->tensorflow) (2024.8.30)\n",
            "Requirement already satisfied: markdown>=2.6.8 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.18,>=2.17->tensorflow) (3.7)\n",
            "Requirement already satisfied: tensorboard-data-server<0.8.0,>=0.7.0 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.18,>=2.17->tensorflow) (0.7.2)\n",
            "Requirement already satisfied: werkzeug>=1.0.1 in /usr/local/lib/python3.10/dist-packages (from tensorboard<2.18,>=2.17->tensorflow) (3.1.3)\n",
            "Requirement already satisfied: MarkupSafe>=2.1.1 in /usr/local/lib/python3.10/dist-packages (from werkzeug>=1.0.1->tensorboard<2.18,>=2.17->tensorflow) (3.0.2)\n",
            "Requirement already satisfied: markdown-it-py>=2.2.0 in /usr/local/lib/python3.10/dist-packages (from rich->keras>=3.2.0->tensorflow) (3.0.0)\n",
            "Requirement already satisfied: pygments<3.0.0,>=2.13.0 in /usr/local/lib/python3.10/dist-packages (from rich->keras>=3.2.0->tensorflow) (2.18.0)\n",
            "Requirement already satisfied: mdurl~=0.1 in /usr/local/lib/python3.10/dist-packages (from markdown-it-py>=2.2.0->rich->keras>=3.2.0->tensorflow) (0.1.2)\n",
            "Downloading biopython-1.84-cp310-cp310-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)\n",
            "\u001b[2K   \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.2/3.2 MB\u001b[0m \u001b[31m57.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hInstalling collected packages: biopython\n",
            "Successfully installed biopython-1.84\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import sys\n",
        "import numpy as np\n",
        "from Bio import SeqIO  # For parsing FASTA files"
      ],
      "metadata": {
        "id": "YUCPLpFjj045"
      },
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Add the path to ProteinBERT code\n",
        "sys.path.append(\"/content/protein_bert\")\n",
        "\n",
        "# Import modules from ProteinBERT\n",
        "from proteinbert.existing_model_loading import load_pretrained_model\n",
        "from proteinbert.tokenization import tokenize_seq, additional_token_to_index\n",
        "from proteinbert.conv_and_global_attention_model import create_model\n",
        "\n",
        "# Step 1: Load the pre-trained ProteinBERT model\n",
        "def load_model(seq_len, vocab_size, n_annotations, model_path):\n",
        "    model_generator, input_encoder = load_pretrained_model(\n",
        "        local_model_dump_dir=\"/content\",\n",
        "        local_model_dump_file_name=model_path\n",
        "    )\n",
        "    model = create_model(seq_len=seq_len, vocab_size=vocab_size, n_annotations=n_annotations)\n",
        "    return model, input_encoder\n",
        "\n",
        "# Step 2: Tokenize the sequences\n",
        "def tokenize_sequences(sequences, seq_len):\n",
        "    tokenized = []\n",
        "    pad_token = additional_token_to_index[\"<PAD>\"]\n",
        "    for seq in sequences:\n",
        "        tokens = tokenize_seq(seq)\n",
        "        # Ensure all sequences are of length `seq_len` (pad or truncate)\n",
        "        if len(tokens) > seq_len:\n",
        "            tokens = tokens[:seq_len]\n",
        "        else:\n",
        "            tokens.extend([pad_token] * (seq_len - len(tokens)))\n",
        "        tokenized.append(tokens)\n",
        "    return np.array(tokenized)\n",
        "\n",
        "# Step 3: Predict sequence probabilities\n",
        "def get_sequence_probabilities(model, tokenized_sequences):\n",
        "    input_annotations = np.zeros((len(tokenized_sequences), model.input_shape[1][1]))  # Dummy annotations\n",
        "    predictions = model.predict([tokenized_sequences, input_annotations])\n",
        "    seq_probabilities = predictions[0]  # Assuming first output corresponds to sequence probabilities\n",
        "    return seq_probabilities\n",
        "\n",
        "# Step 4: Calculate naturalness scores\n",
        "def calculate_naturalness(seq_probs):\n",
        "    # Sum log probabilities as naturalness score for each sequence\n",
        "    return [np.sum(np.log(prob + 1e-8)) for prob in seq_probs]  # Add small epsilon for numerical stability\n",
        "\n",
        "# Main function\n",
        "def find_most_natural_sequence(fasta_path, model_path, seq_len=512, vocab_size=28, n_annotations=0):\n",
        "    model, input_encoder = load_model(seq_len, vocab_size, n_annotations, model_path)\n",
        "\n",
        "    # Parse sequences from FASTA\n",
        "    sequences = [str(record.seq) for record in SeqIO.parse(fasta_path, \"fasta\")]\n",
        "\n",
        "    tokenized_sequences = tokenize_sequences(sequences, seq_len)\n",
        "    seq_probs = get_sequence_probabilities(model, tokenized_sequences)\n",
        "    naturalness_scores = calculate_naturalness(seq_probs)\n",
        "\n",
        "    # Find the sequence with the highest score\n",
        "    most_natural_sequence_idx = np.argmax(naturalness_scores)\n",
        "    most_natural_sequence = sequences[most_natural_sequence_idx]\n",
        "    return most_natural_sequence, naturalness_scores\n",
        "\n",
        "if __name__ == \"__main__\":\n",
        "    # File paths\n",
        "    fasta_path = \"/content/top_10000_lowest_instability_sequences2.fasta\"\n",
        "    model_path = \"/content/epoch_92400_sample_23500000.pkl\"\n",
        "\n",
        "    # Parameters\n",
        "    seq_len = 512  # Adjust based on the model's max sequence length\n",
        "    vocab_size = 28  # 20 amino acids + additional tokens\n",
        "    n_annotations = 0  # Assuming no annotations for naturalness scoring\n",
        "\n",
        "    # Find the most natural sequence\n",
        "    most_natural_sequence, scores = find_most_natural_sequence(fasta_path, model_path, seq_len, vocab_size, n_annotations)\n",
        "    print(\"Most natural sequence:\", most_natural_sequence)\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_0afyBKxiyHX",
        "outputId": "2b27dc34-c6bd-45d8-c05b-17b8221e3a98"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\u001b[1m313/313\u001b[0m \u001b[32m━━━━━━━━━━━━━━━━━━━━\u001b[0m\u001b[37m\u001b[0m \u001b[1m724s\u001b[0m 2s/step\n",
            "Most natural sequence: mmrkksfwfgmltalmlvftmefsvsasaaqpgknvekdyivgfksgvktasikkdiikesggkvdkqfriinagkakldkealkevkndpdvayveedhvghglgqtvpygiplikadkvqaqgfkganvkvavldtgiqashpdlnvvggasfvageayntdgnghgthvagtvaalynttgvlgvapnvslfavkvlnssgngtysgivsgiewvttngmdvinmslggalgstamkqavdhayskgavvvasagnsgssgytntigypakcdsviavgavdsnsnrasfssvgaelevmapgagvystyptntyttlngtsmasphvagtsalilskhpnlsasqvrtrlsstatylgssfsygkglinveaaaq\n"
          ]
        }
      ]
    }
  ]
}